Targeted Sweep Part 1 - Range Tombstones in Cassandra

nziebart opened the pull request:
**Goals (and why)**:
Support Range Tombstones in Cassandra. This is a prerequisite for efficient targeted sweep using a write queue. The intended use case is as follows:
- A write queue contains [Cell, Timestamp] pairs for every committed write
- The background sweeper reads this queue, and for each [Cell c, Timestamp ts] pair, calls `kvs.deleteAllTimestamps(c, ts)`

**Implementation Description (bullets)**:
- Add a generic implementation in `AbstractKeyValueService` that delegates to the existing `delete` method
- For Cassandra, implement the method with range tombstones.

**Concerns (what feedback would you like?)**:
- Is it OK to delete at timestamp `Long.MAX_VALUE` here? We do this for normal deletes, but this is slightly different since we are deleting an entire range rather than specific values. However, I can't see any reason this would not be ok.

**Where should we start reviewing?**:
`CassandraKeyValueServiceImpl.java`

**Priority (whenever / two weeks / yesterday)**:
whenever

<!---
Please remember to:
- Add any necessary release notes (including breaking changes)
- Make sure the documentation is up to date for your change
--->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/2787)
<!-- Reviewable:end -->

nziebart commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java:
```diff
@@ -2069,6 +2072,67 @@ public void deleteRange(final TableReference tableRef, final RangeRequest range)
+
+                @Override
+                public Void apply(CassandraClient client) throws Exception {
+                    Map<ByteBuffer, Map<String, List<Mutation>>> mutationsByRowByTable = Maps.newHashMap();
```
name is wrong, should be `mutationsByTableByRow`

gsheasby commented on atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/KeyValueService.java:
```diff
@@ -373,6 +373,17 @@ void putUnlessExists(@QueryParam("tableRef") TableReference tableRef,
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Idempotent
+    void deleteAllTimestamps(@QueryParam("tableRef") TableReference tableRef,
+            Map<Cell, Long> maxTimestampExclusiveByCell);
```
Is it worth adding a serialisation test for this? (do we ever actually expect someone outside Atlas/Nimbus to hit this?)

nziebart replied:
A serialization test for this endpoint? I don't think anyone uses these endpoints

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java:
```diff
@@ -2069,6 +2072,67 @@ public void deleteRange(final TableReference tableRef, final RangeRequest range)
+                public Void apply(CassandraClient client) throws Exception {
+                    Map<ByteBuffer, Map<String, List<Mutation>>> mutationsByRowByTable = Maps.newHashMap();
+
+                    maxTimestampExclusiveByCell.forEach((cell, maxTimestampExclusive) -> {
```
This method is rather long, and this lambda in particular seems big for a lambda to me. Consider whether pulling out methods would make it more readable (I'd happily pair on this if we weren't sitting so distantly)

nziebart replied:
done

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java:
```diff
@@ -2069,6 +2072,67 @@ public void deleteRange(final TableReference tableRef, final RangeRequest range)
+                        // TODO(nziebart): figure out if/why we can't use the provided max timestamp for the cell here
+                        del.setTimestamp(Long.MAX_VALUE);
+
+                        Mutation mutation = new Mutation();
```
Thought: I'm wondering if we want helper methods that would take a `columnName` and `maxTimestampExclusive`, and give us a "delete range tombstone" mutation.

This would help make this method more readable by encapsulating the boilerplate like "ok, I want to have a Mutation, but first I need a Deletion, and oh jeez, that means I want this slice predicate and I have to hard-code `Long.MAX_VALUE`" - instead, the method becomes "cool, I pass the column name and max timestamp into this helper method and add it to my collection".

nziebart replied:
done

gsheasby commented on atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/thrift/SlicePredicates.java:
```diff
@@ -35,6 +35,14 @@ public static SlicePredicate latestVersionForColumn(byte[] columnName, long maxT
                 Limit.ONE);
     }
 
+    public static SlicePredicate rangeTombstoneForColumn(byte[] columnName, long maxTimestampExclusive) {
```
It would be nice to see some testing around this, but short of actually running this on a test stack and looking with cqlsh at what gets written, I'm not sure how I'd go about it.

nziebart replied:
The ETE tests should give us coverage for this

gsheasby commented on atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/AbstractKeyValueService.java:
```diff
@@ -311,6 +312,33 @@ public void deleteRange(TableReference tableRef, RangeRequest range) {
+        deleteAllTimestampsDefaultImpl(this, tableRef, maxTimestampExclusiveByCell);
+    }
+
+    public static void deleteAllTimestampsDefaultImpl(KeyValueService kvs, TableReference tableRef,
```
We should remember to inline this once JDBC can be removed. Incidentally why couldn't JDBC just use the abstract KVS's implementation?

nziebart replied:
it doesn't extend AbstractKvs - I tried to make it, but it involved some changes, which I didn't think were worth it given this is deprecated

nziebart commented:
@gsheasby addressed your comments. Re testing: the ETE tests are intended to give us coverage here. Happy to add more if you feel they are missing something
